### PR TITLE
lua: Build a shared library

### PR DESCRIPTION
**Pull request by michaelkuhn**
Patch is taken from Arch Linux.

### CODE REVIEW

File: var/spack/repos/builtin/packages/lua/package.py [hunk 1]
```diff
@@ -58,24 +58,31 @@ class Lua(Package):
...
+    patch('https://git.archlinux.org/svntogit/packages.git/plain/trunk/liblua.so.patch?h=packages/lua&id=70abed4cfa74993348bf7afbbc70151a3d57cc01',
+            sha256='2cc83c77423a2dda3696766b2d1ccee2796e052ab04d5178905f41ed9241a3d8')
+
     def install(self, spec, prefix):
         if spec.satisfies("platform=darwin"):
             target = 'macosx'
         else:
             target = 'linux'
         make('INSTALL_TOP=%s' % prefix,
+             'MYCFLAGS=-fPIC',
```
Comment (davydden) [inline var/spack/repos/builtin/packages/lua/package.py]: here and below please use `self.compiler.pic_flag`

Comment (michaelkuhn) [reply var/spack/repos/builtin/packages/lua/package.py]: Done. :-)

File: var/spack/repos/builtin/packages/lua/package.py [hunk 2]
```diff
@@ -58,24 +58,33 @@ class Lua(Package):
...
              target)
         make('INSTALL_TOP=%s' % prefix,
+             'MYCFLAGS=%s' % (
+                 self.compiler.pic_flag),
              'MYLDFLAGS=-L%s -L%s' % (
                  spec['readline'].prefix.lib,
                  spec['ncurses'].prefix.lib),
              'MYLIBS=-lncursesw',
              'CC=%s -std=gnu99' % spack_cc,
+             'TO_LIB=liblua.a liblua.so liblua.so.%s liblua.so.%s' % (
```
Comment (davydden) [inline var/spack/repos/builtin/packages/lua/package.py]: sorry, another issue: this will certainly fail on macOS. Can you please get the list dynamically?

Comment (davydden) [reply var/spack/repos/builtin/packages/lua/package.py]: ah, I see that your patch won't work on macOS either, maybe don't apply it on darwin and skip this line as well?

File: var/spack/repos/builtin/packages/lua/package.py [hunk 3]
```diff
@@ -58,24 +58,33 @@ class Lua(Package):
         destination="luarocks",
         placement='luarocks')
 
+    patch('https://git.archlinux.org/svntogit/packages.git/plain/trunk/liblua.so.patch?h=packages/lua&id=70abed4cfa74993348bf7afbbc70151a3d57cc01',
```
Comment (davydden) [inline var/spack/repos/builtin/packages/lua/package.py]: maybe add this patch to Spack and also add another one from Homebrew: https://github.com/Homebrew/homebrew-core/blob/master/Formula/lua.rb (at the bottom)

File: var/spack/repos/builtin/packages/lua/package.py (hunk 2 above)
Comment (davydden) [reply var/spack/repos/builtin/packages/lua/package.py]: on macOS that would be
```
TO_LIB=liblua.a liblua.dylib
```

**Comment by davydden** [comment]
if you put some `if's` in place, i can check this PR locally on macOS and clean that part

**Comment by michaelkuhn** [comment]
I tried to come up with something a bit more generic. The Linux path still works but you will have to test the macOS path. :-)

File: var/spack/repos/builtin/packages/lua/package.py [hunk 4]
```diff
@@ -58,24 +58,34 @@ class Lua(Package):
         destination="luarocks",
         placement='luarocks')
 
+    patch('liblua-shared.patch')
```
Comment (junghans) [inline var/spack/repos/builtin/packages/lua/package.py]: Add a comment with a link to Arch linux!

Comment (davydden) [reply var/spack/repos/builtin/packages/lua/package.py]: @junghans see the last comment https://github.com/LLNL/spack/pull/6008#issuecomment-340100568

**Comment by junghans** [comment]
Honestly, I would prefer converting the liblua.a to liblua.so outside of their buildsystem.
Calling `gcc -shared -Wl,-soname=liblua.1 -Wl,-z,defs -Wl,--whole-archive liblua.a -Wl,--no-whole-archive -lm -o liblua.so.1` in `install()` doesn't seem too hard.
(Or OSX it would be `gcc -dynamiclib -install_name <prefix>/liblua.1 -Wl,-all_load -Wl,liblua.a -lm -o liblua.so.1`)

@tgamblin: having a generic static_to_shared_lib function would be a good idea anyhow...

**Comment by michaelkuhn** [comment]
I have added a comment with links to the Arch Linux and Homebrew patches. If the conversion function is preferred, though, I could also give that a try in the next few days.

**Pull request merged**